conveyor/generator/generator/parser.go

```diff
@@ -0,0 +1,274 @@
+	p.Package = p.sourceNode.Name.Name
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Parser) findEachStateMachine() {
+	for _, decl := range p.sourceNode.Decls {
+		genDecl, ok := decl.(*ast.GenDecl)
+		if !ok {
+			continue
+		}
+		for _, spec := range genDecl.Specs {
+			currType, ok := spec.(*ast.TypeSpec)
+			if !ok {
+				continue
+			}
+			currStruct, ok := currType.Type.(*ast.InterfaceType)
+			if !ok || !isStateMachineTag(genDecl.Doc) {
+				continue
+			}
+
+			machine := &stateMachine{Name: currType.Name.Name, Module: p.module}
+			p.parseStateMachineInterface(machine, currStruct)
+		}
+	}
+}
+
+func isStateMachineTag(group *ast.CommentGroup) bool {
+	for _, comment := range group.List {
+		if strings.Contains(comment.Text, "conveyer: state_machine") {
+			return true
+		}
+	}
+	return false
+}
+
+func remap(code []byte, m *ast.FieldList) []string {
+	if m == nil {
+		return make([]string, 0)
+	}
+	result := make([]string, len(m.List))
+	for i, z := range m.List {
+		t := z.Type
+		result[i] = string(code[t.Pos()-1 : t.End()-1])
+	}
+	return result
+}
+
+func (p *Parser) parseStateMachineInterface(machine *stateMachine, source *ast.InterfaceType) {
+	var curPos token.Pos = 0
+	for _, methodItem := range source.Methods.List {
+		if len(methodItem.Names) == 0 {
+			continue
+		}
+		if methodItem.Pos() <= curPos {
+			log.Fatal("Incorrect order of methods")
+		}
+		curPos = methodItem.Pos()
+		methodType := methodItem.Type.(*ast.FuncType)
+
+		currentHandler := &handler{
+			Name: methodItem.Names[0].Name,
+			Params: remap(p.sourceCode, methodType.Params),
+			Results: remap(p.sourceCode, methodType.Results),
+		}
+		switch {
+		case currentHandler.Name == "TID":
+		case currentHandler.Name == "Init":
+			p.parseInit(machine, currentHandler)
+		case strings.HasPrefix(currentHandler.Name, "State"):
+			p.parseState(machine, currentHandler)
+		case strings.HasPrefix(currentHandler.Name, "Transit"):
+			p.parseTransit(machine, currentHandler)
+		case strings.HasPrefix(currentHandler.Name, "Migrate"):
+			p.parseMigrate(machine, currentHandler)
+		case strings.HasPrefix(currentHandler.Name, "Error") && len(machine.States) == 0:
+			p.parseInitError(machine, currentHandler)
+		case strings.HasPrefix(currentHandler.Name, "Error"):
+			p.parseError(machine, currentHandler)
+		default:
+			log.Fatal("Unknokn handler: ", currentHandler.Name)
+		}
+	}
+	p.StateMachines = append(p.StateMachines, machine)
+	p.generator.stateMachines = append(p.generator.stateMachines, machine)
+}
+
+func (*Parser) parseInit(machine *stateMachine, h *handler) {
+	if len(h.Params) != 1 {
+		log.Fatal("Init must have only one parameter")
+	}
+	if len(h.Results) != 3 {
+		log.Fatal("Init should return three values")
+	}
+	if !strings.HasPrefix(h.Results[0], "*") {
+		log.Fatal("Returned payload should be a pointer")
+	}
+	if h.Results[1] != "common.ElState" {
+		log.Fatal("Returned state should be common.ElState")
```

Comment:
More clear comment is "Second return parameter must be common.ElState".
For giving more accurate info, add handler name to log.